src/library/itunes/itunesmacosimporter.mm

```diff
@@ -0,0 +1,291 @@
+        if (!isPlaylistShown(playlist)) {
+            return;
+        }
+
+        int playlistId = dbIdFromPersistentId(playlist.persistentID);
+        int parentId = playlist.parentID
+                ? dbIdFromPersistentId(playlist.parentID)
+                : -1;
+
+        QString playlistName = uniquifyPlaylistName(qStringFrom(playlist.name));
+
+        queryInsertToPlaylists.bindValue(":id", playlistId);
+        queryInsertToPlaylists.bindValue(":name", playlistName);
+
+        if (!queryInsertToPlaylists.exec()) {
+            LOG_FAILED_QUERY(queryInsertToPlaylists);
+            return;
+        }
+
+        m_playlistNameByDbId.insert(playlistId, playlistName);
+        m_playlistChildsByDbId.insert(parentId, playlistId);
+
+        int i = 0;
+        for (ITLibMediaItem* item in playlist.items) {
+            if (m_cancelImport) {
+                return;
+            }
+
+            queryInsertToPlaylistTracks.bindValue(":playlist_id", playlistId);
+            queryInsertToPlaylistTracks.bindValue(
+                    ":track_id", dbIdFromPersistentId(item.persistentID));
+            queryInsertToPlaylistTracks.bindValue(":position", i);
+
+            if (!queryInsertToPlaylistTracks.exec()) {
+                LOG_FAILED_QUERY(queryInsertToPlaylistTracks);
+                return;
+            }
+
+            i++;
+        }
+    }
+
+    void importMediaItem(ITLibMediaItem* item, QSqlQuery& query) {
+        query.bindValue(":id", dbIdFromPersistentId(item.persistentID));
+        query.bindValue(":artist", qStringFrom(item.artist.name));
+        query.bindValue(":title", qStringFrom(item.title));
+        query.bindValue(":album", qStringFrom(item.album.title));
+        query.bindValue(":album_artist", qStringFrom(item.album.albumArtist));
+        query.bindValue(":genre", qStringFrom(item.genre));
+        query.bindValue(":grouping", qStringFrom(item.grouping));
+        query.bindValue(":year", static_cast<int>(item.year));
+        query.bindValue(":duration", static_cast<int>(item.totalTime / 1000));
+        query.bindValue(":location", qStringFrom([item.location path]));
+        query.bindValue(":rating", static_cast<int>(item.rating / 20));
+        query.bindValue(":comment", qStringFrom(item.comments));
+        query.bindValue(":tracknumber", static_cast<int>(item.trackNumber));
+        query.bindValue(":bpm", static_cast<int>(item.beatsPerMinute));
+        query.bindValue(":bitrate", static_cast<int>(item.bitrate));
+
+        if (!query.exec()) {
+            LOG_FAILED_QUERY(query);
+            return;
+        }
+    }
+};
+
+} // anonymous namespace
+
+ITunesMacOSImporter::ITunesMacOSImporter(LibraryFeature* parentFeature,
+        QSqlDatabase& database,
+        std::atomic<bool>& cancelImport)
+        : m_parentFeature(parentFeature),
+          m_database(database),
+          m_cancelImport(cancelImport) {
+}
+
+ITunesImport ITunesMacOSImporter::importLibrary() {
+    @autoreleasepool {
```

Comment:
could you explain `@autoreleasepool` to me? From what I could read so far I assume it causes all heap allocations within to be allocated in a memorypool which gets freed at the end of the block?

Reply:
Yes, all Objective-C objects that are allocated within an `@autoreleasepool` block are released (i.e. deallocated) at the end of the block, as long as `autorelease` is called on them during initialization (e.g. `[[[Class alloc] init] autorelease]`.

After reading up on this a bit more, it turns out I was mistaken in that this is sufficient: Usually the compiler inserts this `autorelease` call automatically when ARC (Automatic Reference Counting) is enabled, but after experimenting around with it a bit it seems as if `clang` doesn't automatically enable ARC outside of Xcode projects, contrary to what I had assumed.

Therefore it looks like we have three options to avoid leaking memory:

- Manually insert `release`/`retain` calls (tedious)
- Call `autorelease` on every allocated ObjC object (slightly less tedious)
- Enable ARC in the `CMakeLists.txt` (lets the compiler auto-insert `release`/ `retain` calls)

The latter is the safest and most convenient solution, since we don't need to worry about ObjC memory management at all and can even remove the `@autoreleasepool` since the compiler will automatically make sure that objects are deallocated once the last reference goes out of scope. This is consistent with what Stackoverflow suggests:

- https://stackoverflow.com/questions/26241739/should-i-use-arc-when-coding-in-objective-c

For this reason, I've enabled ARC in e23b3f596df1e6533dfa96b154f1fd4296c04ae7 and removed the `@autoreleasepool`.

---

That `-fobjc-arc` is 'safe' can be verified with the following Objective-C program:

```objc
#import <Foundation/Foundation.h>

@interface Test : NSObject
- (void) dealloc;
@end

@implementation Test
- (void) dealloc {
  NSLog(@"Deallocing!");
}
@end

int main() {
  Test *test = [[Test alloc] init];
  return 0;
}
```

Compiling it with `clang -framework Foundation -fobjc-arc <file name>` causes the compiler to automatically release the object, thus printing `Deallocing`, whereas compiling it with `clang -framework Foundation <file name>` doesn't print anything (leaking the object).

Reply:
Thank you, I agree that adding automatic reference counting is best as long as we take care to avoid reference cycles.